python/ray/tests/test_autoscaling_policy.py

```diff
@@ -0,0 +1,381 @@
+                dynamic_resources=node.available_resources,
+                update_resource_load=False,
+                resource_load={},
+                waiting_bundles=waiting_bundles,
+                infeasible_bundles=infeasible_bundles,
+                pending_placement_groups=[],
+            )
+
+        self.autoscaler.update()
+        self._launch_nodes()
+        self._update_cluster_state()
+
+    def process_event(self, event):
+        if event.event_type == SIMULATOR_EVENT_AUTOSCALER_UPDATE:
+            self.run_autoscaler()
+            next_update = self.virtual_time + self.autoscaler_update_interval_s
+            self.event_queue.put(
+                Event(next_update, SIMULATOR_EVENT_AUTOSCALER_UPDATE))
+        elif event.event_type == SIMULATOR_EVENT_TASK_DONE:
+            task = event.data
+            task.node.free(task.resources)
+            if task.done_callback:
+                task.done_callback()
+        elif event.event_type == SIMULATOR_EVEN_NODE_JOINED:
+            node = event.data
+            node.in_cluster = True
+
+    def step(self):
+        self.virtual_time = self.event_queue.queue[0].time
+        while self.event_queue.queue[0].time == self.virtual_time:
+            event = self.event_queue.get()
+            self.process_event(event)
+        self.schedule()
+        return self.virtual_time
+
+
+SAMPLE_CLUSTER_CONFIG = copy.deepcopy(MULTI_WORKER_CLUSTER)
+SAMPLE_CLUSTER_CONFIG["min_workers"] = 0
+SAMPLE_CLUSTER_CONFIG["max_workers"] = 9999
+SAMPLE_CLUSTER_CONFIG["target_utilization_fraction"] = 1.0
+SAMPLE_CLUSTER_CONFIG["available_node_types"]["m4.16xlarge"][
+    "max_workers"] = 100
+SAMPLE_CLUSTER_CONFIG["available_node_types"]["m4.4xlarge"][
+    "max_workers"] = 10000
+
+
+class AutoscalingPolicyTest(unittest.TestCase):
+    def setUp(self):
+        _NODE_PROVIDERS["mock"] = lambda config: self.create_provider
+        self.provider = None
+        self.tmpdir = tempfile.mkdtemp()
+
+    def tearDown(self):
+        self.provider = None
+        del _NODE_PROVIDERS["mock"]
+        _clear_provider_cache()
+        shutil.rmtree(self.tmpdir)
+        ray.shutdown()
+
+    def waitForNodes(self, expected, comparison=None, tag_filters={}):
+        MAX_ITER = 50
+        for i in range(MAX_ITER):
+            n = len(self.provider.non_terminated_nodes(tag_filters))
+            if comparison is None:
+                comparison = self.assertEqual
+            try:
+                comparison(n, expected)
+                return
+            except Exception:
+                if i == MAX_ITER - 1:
+                    raise
+            time.sleep(0.1)
+
+    def create_provider(self, config, cluster_name):
+        assert self.provider
+        return self.provider
+
+    def write_config(self, config):
+        path = self.tmpdir + "/simple.yaml"
+        with open(path, "w") as f:
+            f.write(yaml.dump(config))
+        return path
```

Comment:
lines 293 to 328 are copy pasted from the autoscaler_test.py. I think if we make a change in autoscaler we might end up having to change it in two places and it might cause some problems later. Can we separate these functions into a base class that both test classes in autoscaler_test.py and here inherit from?

Reply:
This small method happens to be duplicated, but in general these fixtures are not the same, for example the mock provider will likely need to be patched/changed in the future.

Reply:
What prevents us from having a parent class that both of these classes inherit from?